In [39]:
# importing data manipulation libraries
import pandas as pd
import numpy as np
# from ydata_profiling import ProfileReport

# importing visualization libraries
import matplotlib.pyplot as plt
import seaborn as sns

# configure notebook for inline plotting
%matplotlib inline

# config pandas to display more than 20 columns
pd.set_option('display.max_columns',250)

# set grid style 
sns.set_style('darkgrid')

# Data Understanding
The objective in this section is to learn as much as possible about the our match data. Here, we will get a high-level practical understnading of our data and by the end of the process, we should have a clear undertanding of the structure of the dataset, how to clean the data, the target variables and possible modelling techniques.

In [10]:
# read data into pandas dataframe
data = pd.read_csv('./data/Matches.csv', header=0)

In [11]:
# check the shape of dataframe
data.shape

(6460, 215)

The dataset consists of 6,460 observations and 215 features some of which are our target variables.

In [13]:
# check types of columns
matches_df.dtypes.unique()

array([dtype('int64'), dtype('O'), dtype('float64'), dtype('bool')],
      dtype=object)

The result indicates that our dataset consists of categorical (object), boolean and numeric (float and int) features. Before proceeding further, let's confirm that the dtypes match what is expected in the documentation.

### To do (Create table of columns and their definitions) - Use datawrapper and insert link 



In [15]:
# inspect the first few rows
data.head(5)

id  homeID  awayID     season    status  roundID  game_week  \
0  2155     150     108  2016/2017  complete       19          1   
1  2156     145     154  2016/2017  complete       19          1   
2  2157     143     142  2016/2017  complete       19          1   
3  2158     144      92  2016/2017  complete       19          1   
4  2159     147     141  2016/2017  complete       19          1   

   revised_game_week       homeGoals awayGoals  homeGoalCount  awayGoalCount  \
0                 -1  ['45+1', '57']    ['47']              2              1   
1                 -1              []    ['82']              0              1   
2                 -1              []    ['74']              0              1   
3                 -1           ['5']    ['59']              1              1   
4                 -1          ['11']    ['67']              1              1   

   totalGoalCount  team_a_corners  team_b_corners  totalCornerCount  \
0               3               5               3                 8   
1               1               7               4                11   
2               1               3               6                 9   
3               2               5               6                11   
4               2               9               6                15   

   team_a_offsides  team_b_offsides  team_a_yellow_cards  team_b_yellow_cards  \
0                1                0                    2                    2   
1                2                2                    3                    2   
2                0                2                    2                    2   
3                4                0                    0                    0   
4                1                3                    3                    5   

   team_a_red_cards  team_b_red_cards  team_a_shotsOnTarget  \
0                 0                 0                     4   
1                 0                 0                     6   
2                 0                 0                     7   
3                 0                 0                     7   
4                 0                 0                     3   

   team_b_shotsOnTarget  team_a_shotsOffTarget  team_b_shotsOffTarget  \
0                     6                      7                      9   
1                     8                      6                      7   
2                     6                      5                      6   
3                     4                      3                      2   
4                     2                      5                      8   

   team_a_shots  team_b_shots  team_a_fouls  team_b_fouls  team_a_possession  \
0            11            15             7            17                 50   
1            12            15            10            14                 55   
2            12            12            12            14                 54   
3            10             6            10            14                 41   
4             8            10            18            13                 46   

   team_b_possession  refereeID  coach_a_ID  coach_b_ID  \
0                 50      685.0       196.0       160.0   
1                 45      688.0       197.0       198.0   
2                 46      360.0       199.0       200.0   
3                 59      537.0       201.0       156.0   
4                 54      693.0       202.0       203.0   

                        stadium_name stadium_location  team_a_cards_num  \
0                KCOM Stadium (Hull)              NaN                 2   
1                Turf Moor (Burnley)              NaN                 3   
2             Selhurst Park (London)              NaN                 2   
3          Goodison Park (Liverpool)              NaN                 0   
4  Riverside Stadium (Middlesbrough)              NaN                 3   

   team_b_cards_num  odds_ft_1  odds_ft_x  odds_ft_2  odds_ft_over05  \
0                 2     

In [16]:
# inspect the last rows
data.tail()

id  homeID  awayID     season      status  roundID  game_week  \
6455  6689328     143     158  2023/2024  incomplete   100543         38   
6456  6689329     151     223  2023/2024  incomplete   100543         38   
6457  6689330     271     162  2023/2024  incomplete   100543         38   
6458  6689331      93     153  2023/2024  incomplete   100543         38   
6459  6689332     251      92  2023/2024  incomplete   100543         38   

      revised_game_week homeGoals awayGoals  homeGoalCount  awayGoalCount  \
6455                 -1        []        []              0              0   
6456                 -1        []        []              0              0   
6457                 -1        []        []              0              0   
6458                 -1        []        []              0              0   
6459                 -1        []        []              0              0   

      totalGoalCount  team_a_corners  team_b_corners  totalCornerCount  \
6455               0              -1              -1                -1   
6456               0              -1              -1                -1   
6457               0              -1              -1                -1   
6458               0              -1              -1                -1   
6459               0              -1              -1                -1   

      team_a_offsides  team_b_offsides  team_a_yellow_cards  \
6455               -1               -1                   -1   
6456               -1               -1                   -1   
6457               -1               -1                   -1   
6458               -1               -1                   -1   
6459               -1               -1                   -1   

      team_b_yellow_cards  team_a_red_cards  team_b_red_cards  \
6455                   -1                -1                -1   
6456                   -1                -1                -1   
6457                   -1                -1                -1   
6458                   -1                -1                -1   
6459                   -1                -1                -1   

      team_a_shotsOnTarget  team_b_shotsOnTarget  team_a_shotsOffTarget  \
6455                    -1                    -1                     -1   
6456                    -1                    -1                     -1   
6457                    -1                    -1                     -1   
6458                    -1                    -1                     -1   
6459                    -1                    -1                     -1   

      team_b_shotsOffTarget  team_a_shots  team_b_shots  team_a_fouls  \
6455                     -1            -1            -1            -1   
6456                     -1            -1            -1            -1   
6457                     -1            -1            -1            -1   
6458                     -1            -1            -1            -1   
6459                     -1            -1            -1            -1   

      team_b_fouls  team_a_possession  team_b_possession  refereeID  \
6455            -1                 -1                 -1        NaN   
6456            -1                 -1                 -1        NaN   
6457            -1                 -1                 -1        NaN   
6458            -1                 -1                 -1        NaN   
6459            -1                 -1                 -1        NaN   

      coach_a_ID  coach_b_ID                           stadium_name  \
6455        -1.0        -1.0                 Selhurst Park (London)   
6456        -1.0        -1.0                    Anfield (Liverpool)   
6457        -1.0        -1.0  Kenilworth Road (Luton, Bedfordshire)   
6458        -1.0        -1.0            Etihad Stadium (Manchester)   
6459        -1.0        -1.0               Bramall Lane (Sheffield)   

     stadium_location  team_a_cards_num  team_b_cards_num  odds_ft_1  \
6455              NaN                -1                -1   

In [18]:
# inspect random sample of rows
data.sample(5)

id  homeID  awayID     season    status  roundID  game_week  \
89      2244     152     149  2016/2017  complete       19          9   
1181    3336     151     146  2013/2014  complete       22          5   
4527  753995     156     144  2008/2009  complete    53612          0   
3194  751643     221     152  2011/2012  complete    53591         16   
3900  753134     162     151  2009/2010  complete    53608         11   

      revised_game_week                homeGoals     awayGoals  homeGoalCount  \
89                   -1  ['1', '21', '62', '70']            []              4   
1181                 -1                       []        ['53']              0   
4527                 -1                       []  ['48', '71']              0   
3194                 -1                   ['88']        ['59']              1   
3900                 -1       ['24', '73', '87']        ['42']              3   

      awayGoalCount  totalGoalCount  team_a_corners  team_b_corners  \
89                0               4               5               7   
1181              1               1               6               6   
4527              2               2              -1              -1   
3194              1               2              -1              -1   
3900              1               4              -1              -1   

      totalCornerCount  team_a_offsides  team_b_offsides  team_a_yellow_cards  \
89                  12                2                3                    3   
1181                12                3                5                    0   
4527                -1               -1               -1                    3   
3194                -1               -1               -1                   -1   
3900                -1               -1               -1                    1   

      team_b_yellow_cards  team_a_red_cards  team_b_red_cards  \
89                      2                 0                 0   
1181                    1                 0                 0   
4527                    4                 0                 0   
3194                   -1                -1                -1   
3900                    0                 0                 2   

      team_a_shotsOnTarget  team_b_shotsOnTarget  team_a_shotsOffTarget  \
89                       8                     6                      6   
1181                     6                     8                      5   
4527                    -1                    -1                     -1   
3194                    -1                    -1                     -1   
3900                    -1                    -1                     -1   

      team_b_shotsOffTarget  team_a_shots  team_b_shots  team_a_fouls  \
89                       10            14            16             7   
1181                      5            11            13             6   
4527                     -1            -1            -1            -1   
3194                     -1            -1            -1            -1   
3900                     -1            -1            -1            -1   

      team_b_fouls  team_a_possession  team_b_possession  refereeID  \
89              17                 44                 56        NaN   
1181             9                 52                 48        NaN   
4527            -1                 -1                 -1        NaN   
3194            -1                 -1                 -1      537.0   
3900            -1                 -1                 -1      361.0   

      coach_a_ID  coach_b_ID                   stadium_name  \
89         209.0       208.0                Stamford Bridge   
1181        37.0       156.0                        Anfield   
4527      8956.0       206.0  Stadium of Light (Sunderland)   
3194       221.0       252.0             DW Stadium (Wigan)   
3900      3610.0       230.0        Craven Cottage (London)   

             stadium_location  team_a_cards_num  team_b_cards_num  odds_ft_1  \
89 

### Columns of interest:
This were are columns that were identified to contain interesting, problematic or missing data and warranted further inspection.

#### Status Column
The first column of interest is the status colum which indicates the whether a match was completed. Looking at the head, tail and random sample snapshots of the df, the status of a match can be either complete or incomplete. Let's inspect unique values in the column:

In [19]:
# check unique values in the status column
data['status'].unique()

array(['complete', 'incomplete', 'suspended'], dtype=object)

Based on the output above, match status can `complete`, `incomplete`, or `suspended`. The dataframe snapshots, particularly the tail, shows that most of the incomplete matches are from the 2023/2024 season. This makes sense because, the 2023/2024 is ongoing and there are bound to be incomplete matches. For our analysis we will want to drop entries where the match status. 

In [21]:
# Check list of incomplete matches
data[matches_df['status'] == 'incomplete' ].sample(10)

id  homeID  awayID     season      status  roundID  game_week  \
6365  6689238     162      92  2023/2024  incomplete   100543         29   
6319  6689192     223     218  2023/2024  incomplete   100543         24   
6408  6689281     211     223  2023/2024  incomplete   100543         33   
6317  6689190      92     209  2023/2024  incomplete   100543         24   
6333  6689206     209     144  2023/2024  incomplete   100543         26   
6395  6689268     271     148  2023/2024  incomplete   100543         32   
6286  6689159     144     158  2023/2024  incomplete   100543         21   
6435  6689308     143     149  2023/2024  incomplete   100543         36   
6373  6689246     152     145  2023/2024  incomplete   100543         30   
6294  6689167      92     218  2023/2024  incomplete   100543         22   

      revised_game_week homeGoals awayGoals  homeGoalCount  awayGoalCount  \
6365                 -1        []        []              0              0   
6319                 -1        []        []              0              0   
6408                 -1        []        []              0              0   
6317                 -1        []        []              0              0   
6333                 -1        []        []              0              0   
6395                 -1        []        []              0              0   
6286                 -1        []        []              0              0   
6435                 -1        []        []              0              0   
6373                 -1        []        []              0              0   
6294                 -1        []        []              0              0   

      totalGoalCount  team_a_corners  team_b_corners  totalCornerCount  \
6365               0              -1              -1                -1   
6319               0              -1              -1                -1   
6408               0              -1              -1                -1   
6317               0              -1              -1                -1   
6333               0              -1              -1                -1   
6395               0              -1              -1                -1   
6286               0              -1              -1                -1   
6435               0              -1              -1                -1   
6373               0              -1              -1                -1   
6294               0              -1              -1                -1   

      team_a_offsides  team_b_offsides  team_a_yellow_cards  \
6365               -1               -1                   -1   
6319               -1               -1                   -1   
6408               -1               -1                   -1   
6317               -1               -1                   -1   
6333               -1               -1                   -1   
6395               -1               -1                   -1   
6286               -1               -1                   -1   
6435               -1               -1                   -1   
6373               -1               -1                   -1   
6294               -1               -1                   -1   

      team_b_yellow_cards  team_a_red_cards  team_b_red_cards  \
6365                   -1                -1                -1   
6319                   -1                -1                -1   
6408                   -1                -1                -1   
6317                   -1                -1                -1   
6333                   -1                -1                -1   
6395                   -1                -1                -1   
6286                   -1                -1                -1   
6435                   -1                -1                -1   
6373                   -1                -1                -1   
6294                   -1                -1                -1   

      team_a_shotsOnTarget  team_b_shotsOnTarget  team_a_shotsOffTarget  \
6365                    -1

In [22]:
# Check seasons from which matches are incomplete
data[data['status'] == 'incomplete' ]['season'].unique()

array(['2023/2024'], dtype=object)

In [ ]:
# Check list of incomplete matches
data[data['status'] == 'suspended' ]['season'].unique()

The above inspection confirms that all matches with the `incomplete` or `suspended` status are the current season. 

#### game_week and revised_game_week columns
The `revised_game_week` column indicates indicates whether the game-week (the week of the season when the match was played) was revised or not. We observe that for inspected sections of the dataframe, no matches appear to have revised match weeks.

In [23]:
# check for unique values in the 'revised_game_week' column
data['revised_game_week'].unique()

array([-1], dtype=int64)

 -1 is the only unique value in the `revised_game_week` column. Thus this column does not provide any useful information and will be dropped.
 As mentioned before, the `game_week`, indicates the week of the season when the match was played. Typically, the EPL games run from from game-week 1 to game-week 38 (assuming a team plays a single EPL game every week - which is usually the case). 

In [24]:
# check for unique values in the 'game_week' column
data['game_week'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38,  0], dtype=int64)

The above output mostly confirms our that there are typically 38 game weeks in a season. However, there appears to be some entries where the column value is 0.

In [25]:
# check for entries where game_week == 0
data[data['game_week'] == 0]

id  homeID  awayID     season    status  roundID  game_week  \
4180  753648      59     142  2008/2009  complete    53612          0   
4181  753649     153     221  2008/2009  complete    53612          0   
4182  753650     147      92  2008/2009  complete    53612          0   
4183  753651     150     162  2008/2009  complete    53612          0   
4184  753652     144     216  2008/2009  complete    53612          0   
...      ...     ...     ...        ...       ...      ...        ...   
4935  755509     153     158  2007/2008  complete    53621          0   
4936  755510      92     151  2007/2008  complete    53621          0   
4937  755511     156      59  2007/2008  complete    53621          0   
4938  755512     206     216  2007/2008  complete    53621          0   
4939  755513     144     157  2007/2008  complete    53621          0   

      revised_game_week                     homeGoals           awayGoals  \
4180                 -1                         ['4']                  []   
4181                 -1                   ['4', '10']              ['48']   
4182                 -1                  ['71', '87']              ['90']   
4183                 -1                  ['23', '81']               ['8']   
4184                 -1                  ['44', '64']  ['21', '66', '90']   
...                 ...                           ...                 ...   
4935                 -1                   ['8', '88']        ['14', '58']   
4936                 -1                            []        ['69', '74']   
4937                 -1                            []              ['24']   
4938                 -1  ['32', '73', '90+1', '90+3']              ['49']   
4939                 -1            ['28', '70', '82']              ['46']   

      homeGoalCount  awayGoalCount  totalGoalCount  team_a_corners  \
4180              1              0               1              -1   
4181              2              1               3              -1   
4182              2              1               3              -1   
4183              2              1               3              -1   
4184              2              3               5              -1   
...             ...            ...             ...             ...   
4935              2              2               4              -1   
4936              0              2               2              -1   
4937              0              1               1              -1   
4938              4              1               5              -1   
4939              3              1               4              -1   

      team_b_corners  totalCornerCount  team_a_offsides  team_b_offsides  \
4180              -1                -1               -1               -1   
4181              -1                -1               -1               -1   
4182              -1                -1               -1               -1   
4183              -1                -1               -1               -1   
4184              -1                -1               -1               -1   
...              ...               ...              ...              ...   
4935              -1                -1               -1               -1   
4936              -1                -1               -1               -1   
4937              -1                -1               -1               -1   
4938              -1                -1               -1               -1   
4939              -1                -1               -1               -1   

      team_a_yellow_cards  team_b_yellow_cards  team_a_red_cards  \
4180                   -1                   -1                -1   
4181                    2                    1                 0   
4182                    1                    2                 0   
4183                    3                    0                 0   
4184                    2                    2                 0   
...                   ...                  ...

In [26]:
# check seasons where game_week == 0 occur
data[data['game_week'] == 0]['season'].unique()

array(['2008/2009', '2007/2008'], dtype=object)

In [27]:
# check season unique game_week values for the 2008/2009 season
data[data['season'] == '2008/2009']['game_week'].unique()

array([0], dtype=int64)

In [28]:
# check season unique game_week values for the 2007/2008 season
data[data['season'] == '2007/2008']['game_week'].unique()

array([0], dtype=int64)

In [29]:
# check season unique game_week values for the 2009/2010 season
data[data['season'] == '2009/2010']['game_week'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 17, 14, 15, 16,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38], dtype=int64)

Further inspection shows that matches from the `2007/2008` and `2008/2009` seasons have all values in the `game_week` column as 0. `game_week` information seems to have been stardardized starting from `2009/2010` season. If `game_week` data from `2007/2008` and `2008/2009` seasons is to be used during EDA or modelling, it will be important to indicate the correct match-week information.

#### homeGoals and awayGoals columns
This columns consist of arrays that store goal timing data (when goals were scored during the match). Upon closer scrutiny of the dataset, it was observed that the columns `homeGoals` and `homeGoals_timings` contained identical information, as did `awayGoals` and `awayGoals_timings`. In the interest of eliminating redundancy and retaining more descriptive columns, it makes sense to remove the `homeGoals` and `awayGoals` columns.

In [30]:
data[['homeGoals', 'awayGoals', 'homeGoals_timings', 'awayGoals_timings' ]].head()

homeGoals awayGoals homeGoals_timings awayGoals_timings
0  ['45+1', '57']    ['47']    ['45+1', '57']            ['47']
1              []    ['82']                []            ['82']
2              []    ['74']                []            ['74']
3           ['5']    ['59']             ['5']            ['59']
4          ['11']    ['67']            ['11']            ['67']

#### corner & offside columns
Dataframe snapshots (head, tail, sample) showed that some seasons, particularly seasons priror to the `2013/2014` season had missing data (encoded as -1) in the following columns 
- `team_a_corners`
- `team_b_corners`
- `totalCornerCount`
- `team_a_offsides`
- `team_b_offsides`
- `team_a_shotsOnTarget`
- `team_b_shotsOnTarget`
- `team_a_shotsOffTarget`
- `team_b_shotsOffTarget`
- `team_a_shots`
- `team_b_shots`
- `team_a_fouls`
- `team_b_fouls`
- `team_a_possession`
- `team_b_possession`

In [33]:
# group by season and get descriptive stats for columns suspected to contain missing data
missing_data_columns = [
                        'team_a_corners',
                        'team_b_corners',
                        'totalCornerCount',
                        'team_a_offsides',
                        'team_b_offsides',
                        'team_a_shotsOnTarget',
                        'team_b_shotsOnTarget',
                        'team_a_shotsOffTarget',
                        'team_b_shotsOffTarget',
                        'team_a_shots',
                        'team_b_shots',
                        'team_a_fouls',
                        'team_b_fouls',
                        'team_a_possession',
                        'team_b_possession'
                        ]

data.groupby('season')[missing_data_columns].describe()

team_a_corners                                                \
                   count      mean       std  min  25%  50%  75%   max   
season                                                                   
2007/2008          380.0 -1.000000  0.000000 -1.0 -1.0 -1.0 -1.0  -1.0   
2008/2009          380.0 -1.000000  0.000000 -1.0 -1.0 -1.0 -1.0  -1.0   
2009/2010          380.0 -1.000000  0.000000 -1.0 -1.0 -1.0 -1.0  -1.0   
2010/2011          380.0 -1.000000  0.000000 -1.0 -1.0 -1.0 -1.0  -1.0   
2011/2012          380.0 -1.000000  0.000000 -1.0 -1.0 -1.0 -1.0  -1.0   
2012/2013          380.0  6.252632  3.158833  0.0  4.0  6.0  8.0  15.0   
2013/2014          380.0  6.092105  2.962286  0.0  4.0  6.0  8.0  14.0   
2014/2015          380.0  6.042105  3.243555  0.0  4.0  6.0  8.0  18.0   
2015/2016          380.0  5.931579  3.247983  0.0  4.0  5.0  8.0  17.0   
2016/2017          380.0  5.673684  3.085217  0.0  3.0  5.0  8.0  19.0   
2017/2018          380.0  5.750000  2.999230  0.0  4.0  5.0  8.0  18.0   
2018/2019          380.0  5.726316  3.025632  0.0  4.0  5.0  8.0  17.0   
2019/2020          380.0  5.771053  3.212394  0.0  3.0  5.0  8.0  17.0   
2020/2021          380.0  5.536842  3.026917  0.0  3.0  5.0  7.0  16.0   
2021/2022          380.0  5.584211  3.014611  0.0  3.0  5.0  7.0  15.0   
2022/2023          380.0  5.610526  3.071856  0.0  3.0  5.0  8.0  17.0   
2023/2024          380.0  2.647368  4.138993 -1.0 -1.0  1.0  6.0  17.0   

          team_b_corners                                                 \
                   count      mean       std  min   25%  50%  75%   max   
season                                                                    
2007/2008          380.0 -1.000000  0.000000 -1.0 -1.00 -1.0 -1.0  -1.0   
2008/2009          380.0 -1.000000  0.000000 -1.0 -1.00 -1.0 -1.0  -1.0   
2009/2010          380.0 -1.000000  0.000000 -1.0 -1.00 -1.0 -1.0  -1.0   
2010/2011          380.0 -1.000000  0.000000 -1.0 -1.00 -1.0 -1.0  -1.0   
2011/2012          380.0 -1.000000  0.000000 -1.0 -1.00 -1.0 -1.0  -1.0   
2012/2013          380.0  4.792105  2.784511  0.0  3.00  4.0  6.0  17.0   
2013/2014          380.0  4.665789  2.695915  0.0  3.00  4.0  6.0  14.0   
2014/2015          380.0  4.663158  2.568782  0.0  3.00  4.0  6.0  13.0   
2015/2016          380.0  4.876316  2.509710  0.0  3.00  5.0  7.0  14.0   
2016/2017          380.0  4.742105  2.705767  0.0  3.00  4.0  6.0  15.0   
2017/2018          380.0  4.544737  2.769148  0.0  2.75  4.0  6.0  13.0   
2018/2019          380.0  4.557895  2.731487  0.0  2.75  4.0  6.0  14.0   
2019/2020          380.0  4.928947  2.839173  0.0  3.00  5.0  7.0  16.0   
2020/2021          380.0  4.613158  2.669151  0.0  3.00  4.0  6.0  13.0   
2021/2022          380.0  4.818421  2.772115  0.0  3.00  5.0  7.0  14.0   
2022/2023          380.0  4.455263  2.814512  0.0  2.00  4.0  6.0  19.0   
2023/2024          380.0  1.873684  3.397165 -1.0 -1.00  1.0  4.0  13.0   

          totalCornerCount                                                    \
                     count       mean       std  min  25%   50%    75%   max   
season                                                                         
2007/2008            380.0  -1.000000  0.000000 -1.0 -1.0  -1.0  -1.00  -1.0   
2008/2009            380.0  -1.000000  0.000000 -1.0 -1.0  -1.0  -1.00  -1.0   
2009/2010            380.0  -1.000000  0.000000 -1.0 -1.0  -1.0  -1.00  -1.0   
2010/2011            380.0  -1.000000  0.000000 -1.0 -1.0  -1.0  -1.00  -1.0   
2011/2012            380.0  -1.000000  0.000000 -1.0 -1.0  -1.0  -1.00  -1.0   
2012/2013            380.0  11.044737  3.681315  0.0  9.0  11.0  13.25  23.0   
2013/2014            380.0  10.757895  3.522909  2.0  8.0  11.0  13.00  21.0   
2014/2015            380.0  10.705263  3.521135  1.0  8.0  11.0  13.00  23.0   
2015/2016            380.0  10.807895  3.604811  1.0  8.0  10.0  13.00  25.0   
2016/2017            380.0  10.415789  3.442826  2.0  8.0  10.0  13.

The output above confirms that seasons prior to the 2013/2014 season indeed are missing data in the columns identified above. Given that shots, shots on target, and corner kicks are occasions in a football/soccer match when the probility of scoring is high, and therefore probably have significant match outcome prediction power, we will likely retain only data from 2013/2014 season onwards.

#### Referee ID column
The 'refereeID column is a unique identifier for the referee who officiated a match. Dataframe snapshots revealed that the referee column had some missing values:

In [59]:
data[data['refereeID'].isnull()].shape

(2784, 215)

In [57]:
data[data['refereeID'].isnull()]['season'].unique()

array(['2016/2017', '2015/2016', '2014/2015', '2013/2014', '2017/2018',
       '2012/2013', '2008/2009', '2007/2008', '2023/2024'], dtype=object)

Output indicates that 2,784 entries in the refereeID column have null values. Given that Referees are randomly selected and are often assumed to have very little influence on the match outcome (assuming match officiation is fair), we can safely drop the referee column.